osh/word_eval.py

```diff
@@ -684,14 +685,31 @@ def _ApplyTestOp(
                 else:
                     is_falsey = False
 
-            elif case(value_e.BashArray):
-                val = cast(value.BashArray, UP_val)
-                # TODO: allow undefined
-                is_falsey = len(val.strs) == 0
+            elif case(value_e.BashArray, value_e.BashAssoc):
+                if val.tag() == value_e.BashArray:
+                    val = cast(value.BashArray, UP_val)
+                    strs = bash_impl.BashArray_GetValues(val)
+                elif val.tag() == value_e.BashAssoc:
+                    val = cast(value.BashAssoc, UP_val)
+                    strs = bash_impl.BashAssoc_GetValues(val)
+                else:
+                    raise AssertionError()
 
-            elif case(value_e.BashAssoc):
-                val = cast(value.BashAssoc, UP_val)
-                is_falsey = len(val.d) == 0
+                if tok.id in (Id.VTest_ColonHyphen, Id.VTest_ColonEquals,
+                              Id.VTest_ColonQMark, Id.VTest_ColonPlus):
+                    if vsub_state.join_array:
+                        sep_width = len(self.splitter.GetJoinChar())
+                    else:
+                        sep_width = 1  # we use ' ' for a[@]
+
+                    if sep_width == 0:
+                        is_falsey = all(len(s) == 0 for s in strs)
```

Comment:
Hmm, OK. I forgot that `all(...)` doesn't translate in mycpp, and I think you turned it into a translation error recently.

Reply:
[`9145e4f..fba1079`](https://github.com/oils-for-unix/oils/compare/9145e4f395bace2e7927445da6ca985abf9aa78e..fba10790d8e4ac3ab6923780ccb7e5e158cb0bda)